homeassistant/components/media_player/cast.py

```diff
@@ -139,191 +190,441 @@ def _async_create_cast_device(hass, chromecast):
+        # attempt. If the initial connection failed, we would never reach
+        # this code anyway.
+        self._available = True
+        self._cast_status = chromecast.status
+        self._media_status = chromecast.media_controller.status
+        _LOGGER.debug("Connection successful!")
+
+    @callback
+    def _async_disconnect(self):
+        """Disconnect Chromecast object if it is set."""
+        if self._chromecast is None:
+            # Can't disconnect if not connected.
+            return
+        _LOGGER.debug("Disconnecting from previous chromecast socket.")
+        self._available = False
+        self._chromecast.disconnect(blocking=False)
+        # Invalidate some attributes
+        self._chromecast = None
+        self._cast_status = None
+        self._media_status = None
+        self._media_status_received = None
+        self._status_listener.invalidate()
+        self._status_listener = None
+
+    async def async_update(self):
+        """Periodically update the properties.
+
+        Even though we receive callbacks for most state changes, some 3rd party
+        apps don't always send them. Better poll every now and then if the
+        chromecast is active (i.e. an app is running).
+        """
+        if self._chromecast is None or not self._available:
+            # Not connected or not available.
+            return
+
+        if self._chromecast.media_controller.is_active:
+            # We can only update status if the media namespace is active
+            self.hass.async_add_job(
+                self._chromecast.media_controller.update_status)
+
+    # ========== Callbacks ==========
+    def new_cast_status(self, cast_status):
+        """Handle updates of the cast status."""
+        self._cast_status = cast_status
+        self.schedule_update_ha_state()
+
+    def new_media_status(self, media_status):
+        """Handle updates of the media status."""
+        self._media_status = media_status
+        self._media_status_received = dt_util.utcnow()
+        self.schedule_update_ha_state()
+
+    def new_connection_status(self, connection_status):
+        """Handle updates of connection status."""
+        from pychromecast.socket_client import CONNECTION_STATUS_CONNECTED
+
+        new_available = connection_status.status == CONNECTION_STATUS_CONNECTED
+        if new_available != self._available:
+            # Connection status callbacks happen often when disconnected.
+            # Only update state when availability changed to put less pressure
+            # on state machine.
+            _LOGGER.debug("Cast device availability changed: %s",
+                          connection_status.status)
+            self._available = new_available
+            self.schedule_update_ha_state()
+
+    # ========== Service Calls ==========
+    def turn_on(self):
+        """Turn on the cast device."""
+        import pychromecast
+
+        if self._chromecast is None:
```

Comment:
When can this happen? We set the cast info in `async_added_to_hass`.

Reply:
When the chromecast "switched host". This happens for example when a cast audio group changes its "elected leader"/master. Then we first disconnect the old chromecast object and then connect the new one.

In that process, `self._chromecast` is briefly changed to `None` to indicate that the old chromecast object is invalid (any commands on the old object would result in exceptions, as the socket is no longer connected)

Additionally, when starting up, there might be a phase when the chromecast object is not connected yet but the entity is already visible in the frontend. This can be reproduced by manually specifying the cast device by `host:` in `configuration.yaml` and disabling any internet connection of your computer.

Reply:
Shouldn't we set the entity to unavailable if it looses the connection to the chromecast? Then you wouldn't need to guard in all the methods.

It sounds weird that we allow entities to be created without a working connection during setup.

Reply:
> Shouldn't we set the entity to unavailable if it looses the connection to the chromecast? 

Yes, that's also what this PR does. When `self._chromecast` is `None` available will also be `False`.

> Then you wouldn't need to guard in all the methods.

Ok, that's good. I put those checks everywhere to protect for potential race conditions, where the service calls are executed even though we're not connected. If the core can guarantee that these methods will never be called with an unavailable state I can also remove those checks.

> It sounds weird that we allow entities to be created without a working connection during setup.

That's true, and also was the previous behavior. The problem with that was that we would block the entire Home Assistant startup waiting for the socket client to connect. If no connection was available, we would even stop startup indefinitely - not good. I tried looking for a solution to that with `asyncio` for some time, but ultimately I think unavailable is better (and quicker).

I mean if we want to we can also try checking the HTTP API of the chromecast before adding the device, that would fix part of the problem (for standard chromecasts).